langchain-tool-retrieval.ipynb

```diff
@@ -22,329 +45,415 @@
     "from langchain.output_parsers import PydanticOutputParser\n",
     "from langchain.pydantic_v1 import BaseModel, Field\n",
     "from langchain.embeddings import SentenceTransformerEmbeddings\n",
-    "\n",
-    "from utils import map_openbb_collection_to_langchain_tools\n",
-    "\n",
+    "from langchain.output_parsers import RetryWithErrorOutputParser\n",
+    "from langchain.llms import OpenAI\n",
     "\n",
     "os.environ[\"TOKENIZERS_PARALLELISM\"] = \"False\"  # Avoid some warnings from HuggingFace\n",
+    "\n",
     "# Set up OpenAI API key\n",
-    "# os.environ[\"OPENAI_API_KEY\"] = \"\"\n",
-    "# openai.api_key = \"\""
+    "import openai\n",
+    "os.environ[\"OPENAI_API_KEY\"] = \"\"\n",
+    "openai.api_key = \"\"\n",
+    "\n",
+    "# Set up OpenBB Personal Access Token from https://my.openbb.co/app/platform/pat\n",
+    "from openbb import obb\n",
+    "from utils import map_openbb_collection_to_langchain_tools  # provides access to OpenBB Tools\n",
+    "obb.account.login(pat=\"\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Set up OpenBB tools for retrieval\n",
+    "\n",
+    "The following will return all OpenBB tools that we want our agent to have access. This matches the layout architecture defined by OpenBB."
    ]
   },
   {
    "cell_type": "code",
-   "execution_count": 23,
+   "execution_count": 5,
    "metadata": {},
    "outputs": [
     {
      "name": "stdout",
      "output_type": "stream",
      "text": [
-      "25\n"
+      "A total of 27 OpenBB tools has been prepared for function calling\n",
+      "\n",
+      "Processing each OpenBB tool description into a list of docs...\n",
+      "\n",
+      "Create embeddings for each of these OpenBB tool descriptions...\n"
      ]
     }
    ],
    "source": [
-    "# Let's set-up our tool retrieval\n",
-    "fundamental_openbb_tools = map_openbb_collection_to_langchain_tools(\n",
-    "    \"/equity/fundamental\"\n",
-    ")\n",
-    "\n",
-    "peer_comparison_openbb_tools = map_openbb_collection_to_langchain_tools(\n",
-    "    '/equity/compare'\n",
-    ")\n",
-    "\n",
-    "estimate_openbb_tools = map_openbb_collection_to_langchain_tools(\n",
-    "    \"/equity/estimates\"\n",
+    "# can also give a single string, but since we want our agent to have more context we provide access to more functions\n",
+    "# TODO: We can create a \"universal_openbb_tools\" that allows access to all OpenBB functions since we use embeddings\n",
+    "# for the description of each, the context size is not an issue. However, with a bigger vector store, the accuracy may\n",
+    "# take a hit.\n",
```

Comment:
Nit:
```python
# TODO: In future we might implement a  `universal_openbb_tools` collection that allows quick LLM access to all OpenBB functions. We don't run into context size issues doing this, since we rely on embeddings
# for our tools., the context size is not an issue. However, with many tools in a vector store, retrieval may become a challenge.
```